lucene/core/src/java/org/apache/lucene/search/knn/SeededKnnCollectorManager.java

```diff
@@ -0,0 +1,174 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.search.knn;
+
+import java.io.IOException;
+import java.util.Arrays;
+import org.apache.lucene.index.KnnVectorValues;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.BulkScorer;
+import org.apache.lucene.search.CollectionTerminatedException;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.KnnCollector;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.search.TopScoreDocCollector;
+import org.apache.lucene.search.TopScoreDocCollectorManager;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.IOFunction;
+
+/** A {@link KnnCollectorManager} that collects results with a timeout. */
+public class SeededKnnCollectorManager implements KnnCollectorManager {
+  private final KnnCollectorManager delegate;
+  private final Weight seedWeight;
+  private final int k;
+  private final IOFunction<LeafReader, KnnVectorValues> vectorValuesSupplier;
+
+  public SeededKnnCollectorManager(
+      KnnCollectorManager delegate,
+      Weight seedWeight,
+      int k,
+      IOFunction<LeafReader, KnnVectorValues> vectorValuesSupplier) {
+    this.delegate = delegate;
+    this.seedWeight = seedWeight;
+    this.k = k;
+    this.vectorValuesSupplier = vectorValuesSupplier;
+  }
+
+  @Override
+  public KnnCollector newCollector(int visitedLimit, LeafReaderContext ctx) throws IOException {
+    // Execute the seed query
+    TopScoreDocCollector seedCollector =
+        new TopScoreDocCollectorManager(
+                k /* numHits */, null /* after */, Integer.MAX_VALUE /* totalHitsThreshold */)
+            .newCollector();
+    final LeafReader leafReader = ctx.reader();
+    final LeafCollector leafCollector = seedCollector.getLeafCollector(ctx);
+    if (leafCollector != null) {
+      try {
+        BulkScorer scorer = seedWeight.bulkScorer(ctx);
+        if (scorer != null) {
+          scorer.score(
+              leafCollector,
+              leafReader.getLiveDocs(),
+              0 /* min */,
+              DocIdSetIterator.NO_MORE_DOCS /* max */);
+        }
+        leafCollector.finish();
+      } catch (
+          @SuppressWarnings("unused")
+          CollectionTerminatedException e) {
+      }
+    }
+
+    TopDocs seedTopDocs = seedCollector.topDocs();
+    KnnVectorValues vectorValues = vectorValuesSupplier.apply(leafReader);
+    if (seedTopDocs.totalHits.value() == 0 || vectorValues == null) {
+      return delegate.newCollector(visitedLimit, ctx);
+    }
+    KnnVectorValues.DocIndexIterator indexIterator = vectorValues.iterator();
+    DocIdSetIterator seedDocs = new MappedDISI(indexIterator, new TopDocsDISI(seedTopDocs));
+    return new SeededKnnCollector(delegate.newCollector(visitedLimit, ctx), seedDocs);
```

Comment:
subjective: perhaps just one `delegate.newCollector(visitedLimit, ctx)` call would make this code section easier to understand e.g.
```suggestion
    KnnCollector knnCollector = delegate.newCollector(visitedLimit, ctx);
    if (0 < seedTopDocs.totalHits.value()  && vectorValues != null) {
      KnnVectorValues.DocIndexIterator indexIterator = vectorValues.iterator();
      DocIdSetIterator seedDocs = new MappedDISI(indexIterator, new TopDocsDISI(seedTopDocs));
      knnCollector = new SeededKnnCollector(knnCollector, seedDocs);
    }
    return knnCollector;
```